elements/lisk-transactions/src/15_proof_of_misbehavior_transaction.ts

```diff
@@ -0,0 +1,442 @@
+				),
+			);
+		}
+
+		// tslint:disable-next-line no-magic-numbers
+		if (Math.abs(this.asset.header2.height - currentHeight) >= 260000) {
+			errors.push(
+				new TransactionError(
+					'Difference between header2.height and current height must be less than 260000.',
+					this.id,
+					'.asset.header2',
+					this.asset.header2.height,
+				),
+			);
+		}
+
+		/*
+			Check if delegate is eligible to be punished
+		*/
+
+		if (!delegateAccount.isDelegate || !delegateAccount.username) {
+			errors.push(
+				new TransactionError(
+					'Account is not a delegate',
+					this.id,
+					'.asset.header1.generatorPublicKey',
+				),
+			);
+		}
+
+		if (delegateAccount.delegate.isBanned) {
+			errors.push(
+				new TransactionError(
+					'Cannot apply proof-of-misbehavior. Delegate is banned.',
+					this.id,
+					'.asset.header1.generatorPublicKey',
+					this.asset.header1.generatorPublicKey,
+				),
+			);
+		}
+
+		if (
+			isPunished(
+				delegateAccount,
+				delegateAccount,
+				store.chain.lastBlockHeader.height,
+			)
+		) {
+			errors.push(
+				new TransactionError(
+					'Cannot apply proof-of-misbehavior. Delegate is already punished. ',
+					this.id,
+					'.asset.header1.generatorPublicKey',
+					this.asset.header1.generatorPublicKey,
+				),
+			);
+		}
+
+		/* 
+			Check block signatures validity 
+		*/
+
+		const blockHeader1Bytes = Buffer.concat([
+			Buffer.from(networkIdentifier, 'hex'),
+			getBlockBytes(this.asset.header1),
+		]);
+		const blockHeader2Bytes = Buffer.concat([
+			Buffer.from(networkIdentifier, 'hex'),
+			getBlockBytes(this.asset.header2),
+		]);
+
+		const { valid: validHeader1Signature } = validateSignature(
+			this.asset.header1.generatorPublicKey,
+			this.asset.header1.blockSignature,
+			blockHeader1Bytes,
+		);
+
+		if (!validHeader1Signature) {
+			errors.push(
+				new TransactionError(
+					'Invalid block signature for header 1.',
+					this.id,
+					'.asset.header1.blockSignature',
+					this.asset.header1.blockSignature,
+				),
+			);
+		}
+		const { valid: validHeader2Signature } = validateSignature(
+			this.asset.header2.generatorPublicKey,
+			this.asset.header2.blockSignature,
+			blockHeader2Bytes,
+		);
+
+		if (!validHeader2Signature) {
+			errors.push(
+				new TransactionError(
+					'Invalid block signature for header 2.',
+					this.id,
+					'.asset.header2.blockSignature',
+					this.asset.header2.blockSignature,
+				),
+			);
+		}
+
+		/*
+			Update sender account
+		*/
+		const reward =
+			store.chain.lastBlockReward > delegateAccount.balance
+				? delegateAccount.balance
+				: store.chain.lastBlockReward;
+
+		senderAccount.balance += reward;
+		// We store the correct reward value in the asset at apply time to allow for undoing the transaction at a later point in time
+		this.asset.reward = reward;
+
+		store.account.set(senderAccount.address, senderAccount);
+
+		/*
+			Update delegate account
+		*/
+		delegateAccount.delegate.pomHeights.push(currentHeight);
```

Comment:
there is a case where delegateAccount === senderAccount, so before updating the delegate account, it should `get` again after `set` the sender account